README.md
# LooksMapping Scraper

A comprehensive web scraping toolkit for extracting restaurant data from LooksMapping.com, focusing on Manhattan restaurants with demographic and attractiveness metrics.

[![Python 3.8+](https://img.shields.io/badge/python-3.8+-blue.svg)](https://www.python.org/downloads/)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

## 🚀 Features

- **Multiple Scraping Methods**: HTTP requests, Selenium WebDriver, and Playwright automation
- **Comprehensive Data Extraction**: Restaurant names, neighborhoods, attractiveness scores, age demographics, gender ratios
- **Neighborhood Analysis**: Statistical analysis of Manhattan neighborhoods by various metrics
- **Robust Error Handling**: Multiple fallback strategies for data extraction
- **CLI Interface**: Easy-to-use command-line tools for scraping and analysis
- **Type Safety**: Full type hints with mypy strict mode
- **Testing**: Comprehensive test coverage with pytest

## 📋 Requirements

- Python 3.8 or higher
- Chrome browser (for Selenium and Playwright scrapers)
- Internet connection

## 🛠️ Installation

### 1. Clone the Repository

```bash
git clone https://github.com/yourusername/looksmapping-scraper.git
cd looksmapping-scraper
```

### 2. Create Virtual Environment

```bash
python -m venv venv
source venv/bin/activate  # On Windows: venv\Scripts\activate
```

### 3. Install Dependencies

```bash
# Install production dependencies
pip install -e .

# Install development dependencies
pip install -e ".[dev]"
```

### 4. Install Browser Dependencies

```bash
# Install Playwright browsers
playwright install

# Chrome driver is automatically managed by webdriver-manager
```

## 🚀 Quick Start

### Basic Scraping

```bash
# Scrape using HTTP requests (fastest)
python scripts/scrape.py --method http

# Scrape using Selenium (handles dynamic content)
python scripts/scrape.py --method selenium --headless

# Scrape using Playwright (most comprehensive)
python scripts/scrape.py --method playwright
```

### Data Analysis

```bash
# Analyze scraped data
python scripts/analyze.py --input restaurant_data.json

# Get top neighborhoods by attractiveness
python scripts/analyze.py --input data.json --top 5 --metric avg_attractive

# Show correlation analysis
python scripts/analyze.py --input data.json --correlation --summary
```

## 📊 Data Structure

The scraped data is saved in JSON format with the following structure:

```json
[
  {
    "name": "Restaurant Name",
    "hood": "Neighborhood",
    "attractive_score": 8.5,
    "age_score": 7.2,
    "gender_score": 6.8,
    "cuisine": "Italian",
    "score": "8.5/10",
    "reviewers": "150 reviewers",
    "lat": 40.7589,
    "long": -73.9851,
    "source": "looksmapping.com",
    "scraped_at": "2023-11-30T10:00:00Z"
  }
]
```

### Field Descriptions

- **name**: Restaurant name
- **hood**: Manhattan neighborhood
- **attractive_score**: Attractiveness rating (0-10)
- **age_score**: Age demographic score (0-10)
- **gender_score**: Gender ratio score (0-10, lower = more female)
- **cuisine**: Restaurant cuisine type
- **score**: Overall rating
- **reviewers**: Number of reviewers
- **lat/long**: Geographic coordinates
- **source**: Data source
- **scraped_at**: Timestamp of scraping

## 🔧 Configuration

### Environment Variables

Create a `.env` file in the project root:

```bash
# Copy the example file
cp .env.example .env
```

Key configuration options:

```env
# Scraping Configuration
SCRAPER_TIMEOUT=30
SCRAPER_HEADLESS=true
SCRAPER_MAX_RETRIES=3
SCRAPER_DELAY=1.0

# Browser Configuration
BROWSER_WIDTH=1280
BROWSER_HEIGHT=800

# Data Configuration
DATA_OUTPUT_DIR=./data
DATA_FORMAT=json
```

### Command Line Options

```bash
# Scraping options
python scripts/scrape.py --method http --output data.json --format json

# Analysis options
python scripts/analyze.py --input data.json --output analysis.json --top 10
```

## 🧪 Testing

### Run Tests

```bash
# Run all tests
make test

# Run tests with coverage
pytest --cov=src --cov-report=html

# Run specific test file
pytest tests/test_scrapers/test_http_scraper.py -v
```

### Code Quality

```bash
# Format code
make format

# Run linting
make lint

# Type checking
make type-check

# Run all quality checks
make pre-commit
```

## 📈 Analysis Features

### Neighborhood Rankings

The analyzer provides comprehensive neighborhood analysis:

- **Attractiveness Rankings**: Neighborhoods ranked by average attractiveness
- **Age Demographics**: Areas with younger vs. older demographics
- **Gender Ratios**: Neighborhoods with different gender distributions
- **Statistical Analysis**: Mean, median, standard deviation calculations
- **Correlation Analysis**: Relationships between different metrics

### Example Analysis Output

```
=== TOP 10 NEIGHBORHOODS BY ATTRACTIVENESS ===
neighborhood        avg_attractive  restaurant_count
SoHo                9.2            15
Upper East Side     8.9            12
West Village        8.7            18
Tribeca             8.5            8
Chelsea             8.3            10
```

## 🏗️ Architecture

### Project Structure

```
LooksMappingScraper/
├── src/
│   ├── scrapers/          # Scraping implementations
│   ├── analyzers/         # Data analysis tools
│   ├── models/           # Data models
│   └── utils/            # Utility functions
├── tests/                # Test suite
├── scripts/              # CLI scripts
├── docs/                 # Documentation
└── data/                 # Output data
```

### Scraping Methods

1. **HTTP Scraper** (`HttpScraper`)
   - Fast and lightweight
   - Uses requests and BeautifulSoup
   - Limited to static content

2. **Selenium Scraper** (`SeleniumScraper`)
   - Handles dynamic content
   - Uses Chrome WebDriver
   - Good for JavaScript-heavy sites

3. **Playwright Scraper** (`PlaywrightScraper`)
   - Most comprehensive
   - Advanced browser automation
   - Map interaction and popup handling

## 🚨 Troubleshooting

### Common Issues

1. **"ChromeDriver not found"**
   ```bash
   # The project uses webdriver-manager to auto-download drivers
   pip install webdriver-manager
   ```

2. **"Playwright browsers not installed"**
   ```bash
   playwright install
   ```

3. **"No restaurants found"**
   - Check if the website structure has changed
   - Try different scraping methods
   - Verify internet connection

4. **"Timeout errors"**
   - Increase timeout values in configuration
   - Check if the website is accessible

### Debug Mode

Enable debug mode for detailed logging:

```bash
# Set headless=False to see browser actions
python scripts/scrape.py --method selenium --verbose
```

## 🤝 Contributing

1. Fork the repository
2. Create a feature branch: `git checkout -b feature-name`
3. Make your changes
4. Add tests for new functionality
5. Run quality checks: `make pre-commit`
6. Commit your changes: `git commit -m "feat: add new feature"`
7. Push to the branch: `git push origin feature-name`
8. Submit a pull request

### Development Setup

```bash
# Complete development setup
make dev-setup

# Run CI pipeline locally
make ci
```

## 📝 License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.

## ⚖️ Legal Notice

This project is for educational and research purposes. Please respect the website's terms of service and robots.txt file. Use reasonable delays between requests to avoid overloading the server.

## 🆘 Support

For issues and questions:

1. Check the [troubleshooting section](#-troubleshooting)
2. Review the [documentation](docs/)
3. Open an [issue](https://github.com/yourusername/looksmapping-scraper/issues)

## 📊 Changelog

- **v0.1.0** (2023-12-03): Initial release
  - HTTP, Selenium, and Playwright scrapers
  - Neighborhood analysis tools
  - CLI interface
  - Comprehensive testing suite
  - Full documentation

## 🙏 Acknowledgments

- [LooksMapping.com](https://looksmapping.com) for providing the data source
- [Playwright](https://playwright.dev/) for advanced browser automation
- [Selenium](https://selenium.dev/) for WebDriver automation
- [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/) for HTML parsing

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2023-12-03

### Added
- Initial release of LooksMapping Scraper
- HTTP-based scraper using requests and BeautifulSoup
- Selenium WebDriver scraper for dynamic content
- Playwright scraper with advanced browser automation
- Neighborhood analysis tools with statistical calculations
- Command-line interface for scraping and analysis
- Comprehensive test suite with pytest
- Full type hints with mypy strict mode
- GitHub Actions CI/CD pipeline
- Complete documentation with examples
- Configuration management with environment variables
- Data validation and cleaning utilities
- Multiple output formats (JSON, CSV)
- Manhattan neighborhood filtering
- Correlation analysis and summary statistics
- Robust error handling and retry mechanisms
- Browser automation with map interaction
- Data deduplication and validation
- Performance optimizations
- Security scanning and code quality checks

### Features
- **Multiple Scraping Methods**: HTTP, Selenium, and Playwright
- **Comprehensive Data Extraction**: Restaurant names, neighborhoods, scores
- **Statistical Analysis**: Neighborhood rankings and correlations
- **CLI Tools**: Easy-to-use command-line interface
- **Type Safety**: Full type hints and mypy compliance
- **Testing**: Comprehensive test coverage
- **CI/CD**: Automated testing and deployment
- **Documentation**: Complete API docs and examples

### Technical Details
- Python 3.8+ support
- Modern packaging with pyproject.toml
- Environment-based configuration
- Async support for Playwright
- Mock-based testing
- Code quality tools (black, flake8, mypy)
- Security scanning (safety, bandit)
- Coverage reporting
- Pre-commit hooks


src/utils/config.py
"""
Configuration management for LooksMapping Scraper.

This module handles loading and managing configuration settings
from environment variables and configuration files.
"""

import os
from typing import Dict, Any, Optional
from dotenv import load_dotenv
import logging

logger = logging.getLogger(__name__)


class Config:
    """
    Configuration class for LooksMapping Scraper.
    
    This class manages all configuration settings, loading them from
    environment variables with sensible defaults.
    """
    
    def __init__(self, config_file: Optional[str] = None):
        """
        Initialize configuration.
        
        Args:
            config_file: Optional path to configuration file
        """
        # Load environment variables
        if config_file:
            load_dotenv(config_file)
        else:
            load_dotenv()
        
        # Scraping Configuration
        self.scraper_timeout = int(os.getenv("SCRAPER_TIMEOUT", "30"))
        self.scraper_headless = os.getenv("SCRAPER_HEADLESS", "true").lower() == "true"
        self.scraper_max_retries = int(os.getenv("SCRAPER_MAX_RETRIES", "3"))
        self.scraper_delay = float(os.getenv("SCRAPER_DELAY", "1.0"))
        
        # Browser Configuration
        self.browser_width = int(os.getenv("BROWSER_WIDTH", "1280"))
        self.browser_height = int(os.getenv("BROWSER_HEIGHT", "800"))
        self.browser_user_agent = os.getenv(
            "BROWSER_USER_AGENT",
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
        )
        
        # Data Configuration
        self.data_output_dir = os.getenv("DATA_OUTPUT_DIR", "./data")
        self.data_format = os.getenv("DATA_FORMAT", "json")
        self.data_include_html = os.getenv("DATA_INCLUDE_HTML", "false").lower() == "true"
        
        # Logging Configuration
        self.log_level = os.getenv("LOG_LEVEL", "INFO")
        self.log_format = os.getenv(
            "LOG_FORMAT",
            "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
        )
        self.log_file = os.getenv("LOG_FILE", "scraper.log")
        
        # Analysis Configuration
        self.analysis_include_stats = os.getenv("ANALYSIS_INCLUDE_STATS", "true").lower() == "true"
        self.analysis_output_format = os.getenv("ANALYSIS_OUTPUT_FORMAT", "both")
        self.analysis_neighborhood_filter = os.getenv("ANALYSIS_NEIGHBORHOOD_FILTER", "manhattan")
        
        # Development Configuration
        self.debug = os.getenv("DEBUG", "false").lower() == "true"
        self.test_mode = os.getenv("TEST_MODE", "false").lower() == "true"
        self.mock_data = os.getenv("MOCK_DATA", "false").lower() == "true"
        
        # Ensure data directory exists
        os.makedirs(self.data_output_dir, exist_ok=True)
    
    def get_browser_options(self) -> Dict[str, Any]:
        """Get browser configuration options."""
        return {
            "headless": self.scraper_headless,
            "width": self.browser_width,
            "height": self.browser_height,
            "user_agent": self.browser_user_agent,
        }
    
    def get_scraper_options(self) -> Dict[str, Any]:
        """Get scraper configuration options."""
        return {
            "timeout": self.scraper_timeout,
            "max_retries": self.scraper_max_retries,
            "delay": self.scraper_delay,
        }
    
    def get_logging_config(self) -> Dict[str, Any]:
        """Get logging configuration."""
        return {
            "level": self.log_level,
            "format": self.log_format,
            "file": self.log_file,
        }
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert configuration to dictionary."""
        return {
            "scraper_timeout": self.scraper_timeout,
            "scraper_headless": self.scraper_headless,
            "scraper_max_retries": self.scraper_max_retries,
            "scraper_delay": self.scraper_delay,
            "browser_width": self.browser_width,
            "browser_height": self.browser_height,
            "browser_user_agent": self.browser_user_agent,
            "data_output_dir": self.data_output_dir,
            "data_format": self.data_format,
            "data_include_html": self.data_include_html,
            "log_level": self.log_level,
            "log_format": self.log_format,
            "log_file": self.log_file,
            "analysis_include_stats": self.analysis_include_stats,
            "analysis_output_format": self.analysis_output_format,
            "analysis_neighborhood_filter": self.analysis_neighborhood_filter,
            "debug": self.debug,
            "test_mode": self.test_mode,
            "mock_data": self.mock_data,
        }


def load_config(config_file: Optional[str] = None) -> Config:
    """
    Load configuration from environment variables.
    
    Args:
        config_file: Optional path to configuration file
        
    Returns:
        Config: Loaded configuration instance
    """
    return Config(config_file)
